src/systems/navsat/NavSat.cc

```diff
@@ -0,0 +1,266 @@
+}
+
+//////////////////////////////////////////////////
+void NavSat::Implementation::CreateSensors(EntityComponentManager &_ecm)
+{
+  IGN_PROFILE("NavSat::CreateSensors");
+  if (!this->initialized)
+  {
+    _ecm.Each<components::NavSat, components::ParentEntity>(
+      [&](const Entity &_entity,
+          const components::NavSat *_navSat,
+          const components::ParentEntity *_parent)->bool
+        {
+          this->AddSensor(_ecm, _entity, _navSat, _parent);
+          return true;
+        });
+      this->initialized = true;
+  }
+  else
+  {
+    _ecm.EachNew<components::NavSat, components::ParentEntity>(
+      [&](const Entity &_entity,
+          const components::NavSat *_navSat,
+          const components::ParentEntity *_parent)->bool
+        {
+          this->AddSensor(_ecm, _entity, _navSat, _parent);
+          return true;
+      });
+  }
+}
+
+//////////////////////////////////////////////////
+void NavSat::Implementation::Update(const EntityComponentManager &_ecm)
+{
+  IGN_PROFILE("NavSat::Update");
+
+  _ecm.Each<components::NavSat, components::WorldLinearVelocity>(
+    [&](const Entity &_entity,
+        const components::NavSat * /*_navsat*/,
+        const components::WorldLinearVelocity *_worldLinearVel
+        )->bool
+      {
+        auto it = this->entitySensorMap.find(_entity);
+
+        if (it == this->entitySensorMap.end())
+        {
+          ignerr << "Failed to update NavSat sensor entity [" << _entity
+                 << "]. Entity not found." << std::endl;
+          return true;
+        }
+
+        // Position
+        auto latLonEle = sphericalCoordinates(_entity, _ecm);
+        if (!latLonEle)
+        {
+          ignwarn << "Failed to update NavSat sensor enity [" << _entity
+                  << "]. Spherical coordinates not set." << std::endl;
+          return true;
+        }
+
+        it->second->SetLatitude(IGN_DTOR(latLonEle.value().X()));
+        it->second->SetLongitude(IGN_DTOR(latLonEle.value().Y()));
+        it->second->SetAltitude(latLonEle.value().Z());
+
+        // Velocity in ENU frame
+        it->second->SetVelocity(_worldLinearVel->Data());
+
+        return true;
+      });
+}
+
+//////////////////////////////////////////////////
+void NavSat::Implementation::RemoveSensors(const EntityComponentManager &_ecm)
+{
+  IGN_PROFILE("NavSat::RemoveSensors");
+  _ecm.EachRemoved<components::NavSat>(
+    [&](const Entity &_entity,
+        const components::NavSat *)->bool
+      {
+        auto sensorId = this->entitySensorMap.find(_entity);
+        if (sensorId == this->entitySensorMap.end())
+        {
+          ignerr << "Internal error, missing NavSat sensor for entity ["
+                 << _entity << "]" << std::endl;
+          return true;
+        }
+
+        this->entitySensorMap.erase(sensorId);
+
+        return true;
+      });
+}
+
+IGNITION_ADD_PLUGIN(NavSat, System,
+  NavSat::ISystemPreUpdate,
+  NavSat::ISystemPostUpdate
+)
```

Comment:
Noob question: where does the sensor update actually take place? This plugin creates sensors in `PreUpdate` and updates the sensor values in `PostUpdate`, but I don't understand where sensor update computations actually happen. I tried to do a quick search in the gazebo physics system, thinking that all sensors are updated there, but I didn't see anything (maybe I glanced over it in my search though).

Reply:
That's a very good question, which made me realize I was just repeating this existing issue from other sensors: 

* https://github.com/ignitionrobotics/ign-gazebo/issues/797

So I refactored the system in https://github.com/ignitionrobotics/ign-gazebo/issues/797 so that `EachNew` isn't called during `PreUpdate`. This is how it works now:

1. `SdfEntityCreator` creates the sensor entity and fills some components from SDF
2. `PreUpdate`: the first `PreUpdate` doesn't do anything
3. `Update`: `Physics` populates the components created by `SdfEntityCreator`
4. `PostUpdate`: `NavSat` does a few things:
    1. Creates new `ign-sensors` objects for new sensors
    2. Updates all sensors with the data from the components that were filled by physics
    3. Removes sensors that need to be removed
1. The next `PreUpdate` creates new components for the new sensors created during the previous `PostUpdate`, in this case, it's only `SensorTopic`

Let me know what you think, if this is a reasonable flow I can update other sensors in a new PR to start resolving issue #797.

Reply:
The new flow you described looks good to me! Thanks for updating this and also taking the time to address the other sensors in #797.

Reply:
* https://github.com/ignitionrobotics/ign-gazebo/pull/1281